File: tstl-gc-optimized-collections/ts/gc_optimized_map.ts
/** @luaIterator @tupleReturn */
export interface GCOptimizedMapIterable<T> extends Iterable<T> {}

export class GCOptimizedMap<TKey, TValue> {
    get size(): number {
        return this.items.size;
    }

    private items: Map<TKey, TValue>;

    constructor(other?: ReadonlyArray<[TKey, TValue]>) {
        this.items = new Map<TKey, TValue>(other);
    }

    public iterable(): GCOptimizedMapIterable<[TKey, TValue]> {
        return this.items;
    }

    public clear(): void {
        for (const [k, v] of this.items) {
            this.delete(k);
        }
        return;
    }

    public delete(key: TKey): boolean {
        return this.items.delete(key);
    }

    public get(key: TKey): TValue | undefined {
        return this.items.get(key);
    }

    public has(key: TKey): boolean {
        return this.items.has(key);
    }

    public set(key: TKey, value: TValue): GCOptimizedMap<TKey, TValue> {
        this.items.set(key, value);
        return this;
    }
}

File: tstl-gc-optimized-collections/lua/gc_optimized_map.lua
local exports = {}
local GCOptimizedMap = {}
GCOptimizedMap.__index = GCOptimizedMap
GCOptimizedMap.prototype = {}
GCOptimizedMap.prototype.__index = GCOptimizedMap.prototype
GCOptimizedMap.prototype.constructor = GCOptimizedMap

GCOptimizedMap.new = function(...)
    local self = setmetatable({}, GCOptimizedMap.prototype)
    self:__constructor(...)
    return self
end

GCOptimizedMap.prototype.__constructor = function(self, ...)
    self.items = {}
    self.size = 0

    for _, entry in ipairs({...}) do
        self.items[entry[1]] = entry[2]
        self.size = self.size + 1
    end
end

GCOptimizedMap.prototype.clear = function(self)
    for k in pairs(self.items) do
        self.items[k] = nil
    end
    self.size = 0
end

GCOptimizedMap.prototype.delete = function(self, key)
    if self:has(key) then
        self.items[key] = nil
        self.size = self.size - 1
    end
end

GCOptimizedMap.prototype.entries = function(self)
    return pairs(self.items)
end

GCOptimizedMap.prototype.get = function(self, key)
    return self.items[key]
end

GCOptimizedMap.prototype.has = function(self, key)
    return self.items[key] ~= nil
end

GCOptimizedMap.prototype.set = function(self, key, value)
    if not self:has(key) then
        self.size = self.size + 1
    end
    self.items[key] = value
    return this
end

GCOptimizedMap.prototype.size = function(self)
    return self.size
end

exports.GCOptimizedMap = GCOptimizedMap
return exports
